In [379]:
import numpy as np
import pandas as pd
import ast
from collections import Counter

## 1. Load raw data

In [380]:
df_raw = pd.read_csv('../data/raw/books_1.Best_Books_Ever.csv')

## 2. Basic Overview

In [381]:
df_raw.head(10)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
0,2767052-the-hunger-games,The Hunger Games,The Hunger Games #1,Suzanne Collins,4.33,WINNING MEANS FAME AND FORTUNE.LOSING MEANS CE...,English,9780439023481,"['Young Adult', 'Fiction', 'Dystopia', 'Fantas...","['Katniss Everdeen', 'Peeta Mellark', 'Cato (H...",...,NaN,['Locus Award Nominee for Best Young Adult Boo...,6376780,"['3444695', '1921313', '745221', '171994', '93...",96.0,"['District 12, Panem', 'Capitol, Panem', 'Pane...",https://i.gr-assets.com/images/S/compressed.ph...,2993816,30516,5.09
1,2.Harry_Potter_and_the_Order_of_the_Phoenix,Harry Potter and the Order of the Phoenix,Harry Potter #5,"J.K. Rowling, Mary GrandPré (Illustrator)",4.50,There is a door at the end of a silent corrido...,English,9780439358071,"['Fantasy', 'Young Adult', 'Fiction', 'Magic',...","['Sirius Black', 'Draco Malfoy', 'Ron Weasley'...",...,06/21/03,['Bram Stoker Award for Works for Young Reader...,2507623,"['1593642', '637516', '222366', '39573', '14526']",98.0,['Hogwarts School of Witchcraft and Wizardry (...,https://i.gr-assets.com/images/S/compressed.ph...,2632233,26923,7.38
2,2657.To_Kill_a_Mockingbird,To Kill a Mockingbird,To Kill a Mockingbird,Harper Lee,4.28,The unforgettable novel of a childhood in a sl...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Scout Finch', 'Atticus Finch', 'Jem Finch', ...",...,07/11/60,"['Pulitzer Prize for Fiction (1961)', 'Audie A...",4501075,"['2363896', '1333153', '573280', '149952', '80...",95.0,"['Maycomb, Alabama (United States)']",https://i.gr-assets.com/images/S/compressed.ph...,2269402,23328,NaN
3,1885.Pride_and_Prejudice,Pride and Prejudice,NaN,"Jane Austen, Anna Quindlen (Introduction)",4.26,Alternate cover edition of ISBN 9780679783268S...,English,9999999999999,"['Classics', 'Fiction', 'Romance', 'Historical...","['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...",...,01/28/13,[],2998241,"['1617567', '816659', '373311', '113934', '767...",94.0,"['United Kingdom', 'Derbyshire, England (Unite...",https://i.gr-assets.com/images/S/compressed.ph...,1983116,20452,NaN
4,41865.Twilight,Twilight,The Twilight Saga #1,Stephenie Meyer,3.60,About three things I was absolutely positive.\...,English,9780316015844,"['Young Adult', 'Fantasy', 'Romance', 'Vampire...","['Edward Cullen', 'Jacob Black', 'Laurent', 'R...",...,10/05/05,"['Georgia Peach Book Award (2007)', 'Buxtehude...",4964519,"['1751460', '1113682', '1008686', '542017', '5...",78.0,"['Forks, Washington (United States)', 'Phoenix...",https://i.gr-assets.com/images/S/compressed.ph...,1459448,14874,2.1
5,19063.The_Book_Thief,The Book Thief,NaN,Markus Zusak (Goodreads Author),4.37,Librarian's note: An alternate cover edition c...,English,9780375831003,"['Historical Fiction', 'Fiction', 'Young Adult...","['Liesel Meminger', 'Hans Hubermann', 'Rudy St...",...,09/01/05,['National Jewish Book Award for Children’s an...,1834276,"['1048230', '524674', '186297', '48864', '26211']",96.0,"['Molching (Germany)', 'Germany']",https://i.gr-assets.com/images/S/compressed.ph...,1372809,14168,3.8
6,170448.Animal_Farm,Animal Farm,NaN,"George Orwell, Russell Baker (Preface), C.M. W...",3.95,Librarian's note: There is an Alternate Cover ...,English,9780451526342,"['Classics', 'Fiction', 'Dystopia', 'Fantasy',...","['Snowball', 'Napoleon', 'Clover', 'Boxer', 'O...",...,08/17/45,"['Prometheus Hall of Fame Award (2011)', 'Retr...",2740713,"['986764', '958699', '545475', '165093', '84682']",91.0,"['England', 'United Kingdom']",https://i.gr-assets.com/images/S/compressed.ph...,1276599,13264,4.42
7,11127.The_Chronicles_of_Narnia,The Chronicles of Narnia,The Chronicles of Narnia (Publication Order) #1–7,"C.S. Lewis, Pauline Baynes (Illustrator)",4.26,"Journeys to the end of the world, fantastic cr...",English,9999999999999,"['Fantasy', 'Classics', 'Fiction', 'Youn

## 3. Summary Statistics

In [382]:
df_raw.shape

(52478, 25)

### 3.1 Numeric column summary

In [383]:
df_raw.describe()

,rating,numRatings,likedPercent,bbeScore,bbeVotes
count,52478.000000,5.247800e+04,51856.000000,5.247800e+04,52478.000000
mean,4.021878,1.787865e+04,92.231545,1.984023e+03,22.529003
std,0.367146,1.039448e+05,5.990689,3.515314e+04,369.158541
min,0.000000,0.000000e+00,0.000000,0.000000e+00,-4.000000
25%,3.820000,3.410000e+02,90.000000,8.400000e+01,1.000000
50%,4.030000,2.307000e+03,94.000000,9.700000e+01,1.000000
75%,4.230000,9.380500e+03,96.000000,1.870000e+02,2.000000
max,5.000000,7.048471e+06,100.000000,2.993816e+06,30516.000000


Initial observations:
- Numeric columns include rating, number of ratings, liked percentage, BBE score, and BBE votes.
- `rating` appears to be bounded between 0 and 5.
- `rating` appears roughly centered around 4.0, while popularity-related columns such as `numRatings`, `bbeScore`, and `bbeVotes` are strongly right-skewed.
- For `numRatings`, `bbeScore`, and `bbeVotes`, skew is mainly because a small number of books are extremely popular, while most books have much lower engagement.
- For `likedPercent`, a high concentration near the upper end may reflect selection bias or Goodreads/list bias.
- `bbeVotes` (Number of votes in Best Books Ever list) has impossible value of -4, which need further investigation before analysis.
- `numRatings`, `bbeScore`, and `bbeVotes` show strong right-skew and potential outliers because their maximum values are much larger than their 75th percentile values.
- Minimum values of 0 could be due to missing or placeholder data as the list is voted on by the general Goodreads community.

### 3.2 Text/object column summary

In [384]:
df_raw.describe(include="str")

,bookId,title,series,author,description,language,isbn,genres,characters,bookFormat,edition,pages,publisher,publishDate,firstPublishDate,awards,ratingsByStars,setting,coverImg,price
count,52478,52478,23470,52478,51140,48672,52478,52478,52478,51005,4955,50131,48782,51598,31152,52478,52478,52478,51873,38113
unique,52424,49927,22802,28227,50888,81,48072,44154,12448,136,1788,1365,11110,10635,8024,9215,49908,4651,51819,3776
top,8794263-promises-to-keep,Legacy,Star Wars Legends,Nora Roberts (Goodreads Author),"هذه هي طبعة ""دار الفكر - بيروت"" وهي آخر طبعة ع...",English,9999999999999,[],[],Paperback,First Edition,320,Vintage,Published,01/01/12,[],[],[],https://i.gr-assets.com/images/S/compressed.ph...,5.20
freq,2,14,15,86,37,42661,4354,4623,38712,26773,458,1049,572,586,225,41864,1423,40900,2,626


| Column           | Current type  | Real meaning                   |
| ---------------- | ------------- | ------------------------------ |
|`author`          | object/string | mixed author/contributor field |
|`language`        | object/string | categorical                    |
|`genres`          | object/string | list-like                      |
|`characters`      | object/string | list-like                      |
|`bookFormat`      | object/string | categorical                    |
|`pages`           | object/string | numeric-like                   |
|`publishDate`     | object/string | date-like                      |
|`firstPublishDate`| object/string | date-like                      |
|`awards`          | object/string | list-like                      |
|`ratingsByStars`  | object/string | list-like numeric counts       |
|`setting`         | object/string | list-like                      |
|`price`           | object/string | numeric-like                   |





Initial observations:
- Many object columns are not clean categorical variables.
- Some columns contain list-like strings, such as `genres`, `characters`, `awards`, `ratingsByStars`, and `setting`.
- Some columns are numeric-like or date-like but currently stored as text, such as `pages`, `price`, `publishDate`, and `firstPublishDate`.
- These columns need dedicated cleaning decisions before EDA.

### 3.3 Column role classification
| Role                    |Columns                             | Notes                      |
| ----------------------- | ---------------------------------- | ---------------------------|
| Identifiers             | `bookId`, `isbn`                                                                 | Useful for record tracking, uniqueness checks, and duplicate investigation. Not analytics features without validation. |
| Core metadata           | `title`, `author`, `language`, `bookFormat`, `publisher`, `series`, `edition`    | Useful for filtering, grouping, and spotting duplicates. Some values need cleaning. |
| Reader-response metrics | `rating`, `numRatings`, `likedPercent`, `ratingsByStars`, `bbeScore`, `bbeVotes` | Central to book success, but they measure different signals and should be used together carefully. |
| Temporal fields         | `publishDate`,`firstPublishDate`                                                 | Stored as text with inconsistent formats. Extracting year is more reliable than using full dates. |
| List-like fields        | `genres`, `characters`, `awards`,`setting`                                       | Text fields representing lists; parse them before analysis. |
| Text/NLP field          |`description`                                                                     | Main NLP input, but it reflects book content/marketing, not reader sentiment. |
| Numeric-like attributes | `pages`,`price`                                                                  | Numeric values stored as text. |
| Media/link metadata     |`coverImg`                                                                        | Useful for dashboard display, not core analysis. |


## 4. Schema Inspection

In [385]:
df_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 52478 entries, 0 to 52477
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   bookId            52478 non-null  str    
 1   title             52478 non-null  str    
 2   series            23470 non-null  str    
 3   author            52478 non-null  str    
 4   rating            52478 non-null  float64
 5   description       51140 non-null  str    
 6   language          48672 non-null  str    
 7   isbn              52478 non-null  str    
 8   genres            52478 non-null  str    
 9   characters        52478 non-null  str    
 10  bookFormat        51005 non-null  str    
 11  edition           4955 non-null   str    
 12  pages             50131 non-null  str    
 13  publisher         48782 non-null  str    
 14  publishDate       51598 non-null  str    
 15  firstPublishDate  31152 non-null  str    
 16  awards            52478 non-null  str    
 17  numR

The dataset contains 52478 rows and 25 columns. 20 columns are currently stored as string but may contain categorical, list or numeric values and 5 are numeric. Book metadata columns like title, series are high-cardinality columns. `bookId` is the ID column of the dataset.

## 5. Missingness audit

In [386]:
missing = df_raw.isnull().mean().sort_values(ascending=False).mul(100).round(2)
missing

edition             90.56
series              55.28
firstPublishDate    40.64
price               27.37
language             7.25
publisher            7.04
pages                4.47
bookFormat           2.81
description          2.55
publishDate          1.68
likedPercent         1.19
coverImg             1.15
isbn                 0.00
ratingsByStars       0.00
bbeVotes             0.00
bbeScore             0.00
setting              0.00
awards               0.00
numRatings           0.00
genres               0.00
author               0.00
rating               0.00
title                0.00
characters           0.00
bookId               0.00
dtype: float64

The highest missingness appears in `edition`, `series`, `firstPublishDate`, and `price`.

These missing values should not be handled the same way:
- `edition` is likely optional metadata and may not be useful for core analysis.
- `series` missingness may be meaningful because many books are standalone.
- `firstPublishDate` is important for temporal analysis but has substantial missingness.
- `price` needs further investigation before deciding whether to use it.

## 6. Duplicate Audit

### 6.1 Duplicate type 1: exact full-row duplicates

In [387]:
exact_duplicate_count = df_raw.duplicated().sum()
exact_duplicate_count

np.int64(50)

In [388]:
exact_duplicates = df_raw[df_raw.duplicated(keep=False)].sort_values("bookId")
exact_duplicates.head(10)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
37460,11737313-three-times-lucky,Three Times Lucky,Mo & Dale Mysteries #1,Sheila Turnage,4.02,"Newbery honor winner, New York Times bestselle...",English,9780803736702,"['Mystery', 'Middle Grade', 'Fiction', 'Realis...",[],...,NaN,"['Newbery Medal Nominee (2013)', ""Pennsylvania...",22422,"['8633', '7941', '4181', '1076', '591']",93.0,['North Carolina (United States)'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,3.06
37406,11737313-three-times-lucky,Three Times Lucky,Mo & Dale Mysteries #1,Sheila Turnage,4.02,"Newbery honor winner, New York Times bestselle...",English,9780803736702,"['Mystery', 'Middle Grade', 'Fiction', 'Realis...",[],...,NaN,"['Newbery Medal Nominee (2013)', ""Pennsylvania...",22422,"['8633', '7941', '4181', '1076', '591']",93.0,['North Carolina (United States)'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,3.06
37442,1258333.Cantarella_Volume_1,"Cantarella, Volume 1",Cantarella #1,You Higuri,3.81,"From birth, Cesare Borgia is surrounded by sha...",English,9780976895701,"['Manga', 'Historical Fiction', 'Fantasy', 'Gr...","['Cesare Borgia', 'Lucrezia Borgia', 'Pope Ale...",...,03/08/01,[],933,"['324', '250', '253', '71', '35']",89.0,['Italy'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.67
37388,1258333.Cantarella_Volume_1,"Cantarella, Volume 1",Cantarella #1,You Higuri,3.81,"From birth, Cesare Borgia is surrounded by sha...",English,9780976895701,"['Manga', 'Historical Fiction', 'Fantasy', 'Gr...","['Cesare Borgia', 'Lucrezia Borgia', 'Pope Ale...",...,03/08/01,[],933,"['324', '250', '253', '71', '35']",89.0,['Italy'],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.67
37428,13571757-1,アオハライド 1,アオハライド [Ao Haru Ride] #1,"Io Sakisaka, 咲坂伊緒",4.33,高1の双葉には中学時代から忘れられない初恋の人、「田中くん」がいる。中学のとき女子に、はぶら...,Japanese,9784088466477,"['Manga', 'Romance', 'Young Adult', 'Shojo', '...",[],...,NaN,[],16361,"['9121', '4447', '2102', '451', '240']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,13.30
37482,13571757-1,アオハライド 1,アオハライド [Ao Haru Ride] #1,"Io Sakisaka, 咲坂伊緒",4.33,高1の双葉には中学時代から忘れられない初恋の人、「田中くん」がいる。中学のとき女子に、はぶら...,Japanese,9784088466477,"['Manga', 'Romance', 'Young Adult', 'Shojo', '...",[],...,NaN,[],16361,"['9121', '4447', '2102', '451', '240']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,13.30
37461,13614864-know-not-why,Know Not Why,Know Not Why #1,Hannah Johnson (Goodreads Author),4.02,Howie gets a job at Artie Kraft's Arts 'N Craf...,English,9781476149035,"['M M Romance', 'LGBT', 'Romance', 'Contempora...","[""Howard 'Howie' Jenkins"", 'Arthur Kraft Jr.',...",...,NaN,[],3411,"['1290', '1231', '632', '189', '69']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37407,13614864-know-not-why,Know Not Why,Know Not Why #1,Hannah Johnson (Goodreads Author),4.02,Howie gets a job at Artie Kraft's Arts 'N Craf...,English,9781476149035,"['M M Romance', 'LGBT', 'Romance', 'Contempora...","[""Howard 'Howie' Jenkins"", 'Arthur Kraft Jr.',...",...,NaN,[],3411,"['1290', '1231', '632', '189', '69']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37483,1481056.Gifts_and_Ministries_of_the_Holy_Spirit,Gifts and Ministries of the Holy Spirit,NaN,Lester Sumrall,4.44,You Can Have the Power of the Holy Spirit The...,English,9780883686522,"['Christian', 'Theology', 'Christian Non Ficti...",[],...,06/06/77,[],333,"['206', '83', '31', '11', '2']",96.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,5.00
37429,1481056.Gifts_and_Ministries_of_the_Holy_Spirit,Gifts and Ministries of the Holy Spirit,NaN,Lester Sumrall,4.44,You Can Have the Power of the Holy Spirit The...,English,9780883686522,"['Christian', 'Theology', 'Christian Non Ficti...",[],...,06/06/77,[],333,"['206', '83', '31', '11', '2']",96.0,[],https://i.gr-assets.com/images/S/

Exact duplicate rows likely represent repeated scraped records from Goodreads list. There are 50 exact duplicates. Since all values are identical, removing them should not lose unique information.

### 6.2 Duplicate type 2: duplicate `bookId`

In [389]:
bookid_duplicate_count = df_raw.duplicated(subset=["bookId"]).sum()
bookid_duplicate_count

np.int64(54)

Duplicate `bookId` values are suspicious because `bookId` should identify a unique source record. These records require inspection to determine whether they are exact duplicates, near-duplicates, or inconsistent entries. An overview of the data seems to show that they are exact duplicates most of the time.

### 6.3 Duplicate type 3: duplicate `isbn`

In [390]:
isbn_duplicate = df_raw[ df_raw.duplicated(subset=["isbn"], keep=False)]
isbn_duplicate.head(10)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
2,2657.To_Kill_a_Mockingbird,To Kill a Mockingbird,To Kill a Mockingbird,Harper Lee,4.28,The unforgettable novel of a childhood in a sl...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Scout Finch', 'Atticus Finch', 'Jem Finch', ...",...,07/11/60,"['Pulitzer Prize for Fiction (1961)', 'Audie A...",4501075,"['2363896', '1333153', '573280', '149952', '80...",95.0,"['Maycomb, Alabama (United States)']",https://i.gr-assets.com/images/S/compressed.ph...,2269402,23328,NaN
3,1885.Pride_and_Prejudice,Pride and Prejudice,NaN,"Jane Austen, Anna Quindlen (Introduction)",4.26,Alternate cover edition of ISBN 9780679783268S...,English,9999999999999,"['Classics', 'Fiction', 'Romance', 'Historical...","['Mr. Bennet', 'Mrs. Bennet', 'Jane Bennet', '...",...,01/28/13,[],2998241,"['1617567', '816659', '373311', '113934', '767...",94.0,"['United Kingdom', 'Derbyshire, England (Unite...",https://i.gr-assets.com/images/S/compressed.ph...,1983116,20452,NaN
7,11127.The_Chronicles_of_Narnia,The Chronicles of Narnia,The Chronicles of Narnia (Publication Order) #1–7,"C.S. Lewis, Pauline Baynes (Illustrator)",4.26,"Journeys to the end of the world, fantastic cr...",English,9999999999999,"['Fantasy', 'Classics', 'Fiction', 'Young Adul...","['Polly', 'Aslan', 'Lucy Pevensie', 'Edmund Pe...",...,10/28/56,[],517740,"['254964', '167572', '74362', '15423', '5419']",96.0,"['London, England']",https://i.gr-assets.com/images/S/compressed.ph...,1238556,12949,NaN
10,11870085-the-fault-in-our-stars,The Fault in Our Stars,NaN,John Green (Goodreads Author),4.21,Despite the tumor-shrinking medical miracle th...,English,9999999999999,"['Young Adult', 'Romance', 'Fiction', 'Contemp...","['Hazel Grace Lancaster', 'Augustus Waters', '...",...,NaN,"['Georgia Peach Book Award (2013)', 'Buxtehude...",3550714,"['1784471', '1022406', '512574', '150365', '80...",93.0,"['Indianapolis, Indiana (United States)', 'Ams...",https://i.gr-assets.com/images/S/compressed.ph...,1087056,11287,NaN
11,386162.The_Hitchhiker_s_Guide_to_the_Galaxy,The Hitchhiker's Guide to the Galaxy,The Hitchhiker's Guide to the Galaxy #1,Douglas Adams,4.22,Seconds before the Earth is demolished to make...,English,9999999999999,"['Science Fiction', 'Fiction', 'Humor', 'Fanta...","['Zaphod Beeblebrox', 'Arthur Dent', 'Ford Pre...",...,10/12/79,[],1436325,"['725771', '420864', '199846', '58326', '31518']",94.0,[],https://i.gr-assets.com/images/S/compressed.ph...,1063601,10996,NaN
14,968.The_Da_Vinci_Code,The Da Vinci Code,Robert Langdon #2,Dan Brown (Goodreads Author),3.86,ISBN 9780307277671 moved to this edition.While...,English,9999999999999,"['Fiction', 'Mystery', 'Thriller', 'Suspense',...","['Sophie Neveu', 'Robert Langdon', 'Sir Leigh ...",...,03/18/03,['British Book Award for Book of the Year (200...,1933446,"['645308', '667657', '399278', '142103', '79100']",89.0,"['Paris (France)', 'London, England', 'France'...",https://i.gr-assets.com/images/S/compressed.ph...,876633,9231,NaN
16,5297.The_Picture_of_Dorian_Gray,The Picture of Dorian Gray,NaN,"Oscar Wilde, Jeffrey Eugenides (Introduction)",4.08,"Written in his distinctively dazzling manner, ...",English,9999999999999,"['Classics', 'Fiction', 'Horror', 'Fantasy', '...","['Basil Hallward', 'Lord Henry Wotton', 'Sibyl...",...,07/20/90,[],966196,"['382985', '350778', '176188', '41869', '14376']",94.0,"['England', 'England']",https://i.gr-assets.com/images/S/compressed.ph...,861799,9096,NaN
19,24280.Les_Mis_rables,Les Misérables,NaN,"Victor Hugo, Lee Fahnestock (Translator), Norm...",4.18,Introducing one of the most famous characters ...,English,9999999999999,"['Classics', 'Fiction', 'Historical Fiction', ...","['Jean Valjean', 'Javert', 'Cosette', 'Fantine...",...,10/28/62,[],682793,"['332692', '202012', '100932', '29673', '17484']",93.0,['Pari

Duplicate ISBN values may indicate duplicate edition records, but ISBN analysis should be inspected carefully because some ISBN values are missing, placeholder-like, or malformed.

ISBN duplicate checks are preliminary because ISBN cleaning has not been done yet.

### 6.4 Duplicate type 4: duplicate `title + author`

In [391]:
title_author_duplicates = df_raw[ df_raw.duplicated(subset=["title", "author"], keep=False) ].sort_values(["title", "author"])

title_author_duplicates.head(20)

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
2876,6443834-1q84,1Q84,1Q84 #1,"Haruki Murakami, 村上 春樹",4.00,1949年にジョージ・オーウェルは、近未来小説としての『1984』を刊行した。\nそして20...,Japanese,9784103534228,"['Fiction', 'Fantasy', 'Japan', 'Japanese Lite...",[],...,NaN,[],49121,"['17854', '18852', '8553', '2421', '1441']",92.0,['Tokyo (Japan)'],https://i.gr-assets.com/images/S/compressed.ph...,1218,15,9.09
4463,7717482-1q84,1Q84,1Q84 #3,"Haruki Murakami, 村上 春樹",3.90,1949年にジョージ・オーウェルは、近未来小説としての『1984』を刊行した。\n\nそして...,Japanese,9784103534259,"['Fiction', 'Fantasy', 'Japan', 'Japanese Lite...",[],...,10/28/10,[],28602,"['9232', '10355', '6565', '1952', '498']",91.0,['Tokyo (Japan)'],https://i.gr-assets.com/images/S/compressed.ph...,682,9,13.93
1679,12177850-a-song-of-ice-and-fire,A Song of Ice and Fire,A Song of Ice and Fire #1-5,George R.R. Martin,4.56,"For the first time, all five novels in the epi...",English,9781780484259,"['Fantasy', 'Fiction', 'Science Fiction Fantas...","['Tyrion Lannister', 'Arya Stark', 'Khal Drogo...",...,10/28/00,[],53233,"['36854', '11677', '3118', '890', '694']",97.0,[],https://i.gr-assets.com/images/S/compressed.ph...,2616,35,NaN
2899,9814682-a-song-of-ice-and-fire,A Song of Ice and Fire,A Song of Ice and Fire #1-4,George R.R. Martin,4.63,George R. R. Martin's A Song of Ice and Fire s...,English,9780345529053,"['Fantasy', 'Fiction', 'Epic Fantasy', 'Advent...",[],...,01/01/11,[],44894,"['32256', '9608', '2253', '506', '271']",98.0,[],https://i.gr-assets.com/images/S/compressed.ph...,1201,16,12.24
37383,24633605-always-and-forever,Always and Forever,Serenity Point #2,Harper Bentley (Goodreads Author),3.93,Does wanting to slap the hell out of Brody Kel...,English,9999999999999,"['Contemporary Romance', 'Romance', 'Firefight...",[],...,NaN,[],482,"['153', '188', '104', '26', '11']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37437,24633605-always-and-forever,Always and Forever,Serenity Point #2,Harper Bentley (Goodreads Author),3.93,Does wanting to slap the hell out of Brody Kel...,English,9999999999999,"['Contemporary Romance', 'Romance', 'Firefight...",[],...,NaN,[],482,"['153', '188', '104', '26', '11']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37412,3973924-bal-masqu,Bal masqué,NaN,"Elia Barceló, Dorotea ter Horst (Translator)",3.79,"Parijs, 1991. Jaren nadat de Argentijnse schri...",Dutch,9789021453293,"['Roman', 'Fiction']",[],...,01/01/04,[],341,"['77', '146', '91', '24', '3']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37466,3973924-bal-masqu,Bal masqué,NaN,"Elia Barceló, Dorotea ter Horst (Translator)",3.79,"Parijs, 1991. Jaren nadat de Argentijnse schri...",Dutch,9789021453293,"['Roman', 'Fiction']",[],...,01/01/04,[],341,"['77', '146', '91', '24', '3']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,NaN
37426,26845691-beatlebone,Beatlebone,NaN,Kevin Barry,3.47,He will spend three days alone on his island. ...,English,9781782116134,"['Fiction', 'Ireland', 'Irish Literature', 'Mu...",[],...,NaN,['Goldsmiths Prize (2015)'],2528,"['471', '860', '729', '334', '134']",81.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.16
37480,26845691-beatlebone,Beatlebone,NaN,Kevin Barry,3.47,He will spend three days alone on his island. ...,English,9781782116134,"['Fiction', 'Ireland', 'Irish Literature', 'Mu...",[],...,NaN,['Goldsmiths Prize (2015)'],2528,"['471', '860', '729', '334', '134']",81.0,[],https://i.gr-assets.com/images/S/compressed.ph...,87,1,4.16


In [392]:
len(title_author_duplicates)

171

Duplicate `title + author` combinations should not be automatically removed. They may represent different editions, formats, translations, or publication records and the number might change after removing the exact duplicates from the dataset. These rows are important for understanding the dataset grain.

## 7. Dataset Grain and Duplicate Strategy

Before cleaning or removing duplicate records, I need to define what one row represents. In this dataset, one row appears to represent a Goodreads book record or edition-level entry, not necessarily one unique literary work. This distinction matters because the same book title and author may appear multiple times due to different editions, formats, translations, publication dates, or ISBNs.

| Possible grain     | Meaning                                                | Consequence                                 |
| ------------------ | ------------------------------------------------------ | ------------------------------------------- |
| Record-level       | Each row is one scraped Goodreads record               | Safest assumption for now                   |
| Edition-level      | Each row is a specific edition or format of a book     | Same title/author can appear multiple times |
| Work-level         | Each row is one unique book/work regardless of edition | Would require merging editions              |
| Author-title-level | Each row is one title-author combination               | Too simplistic and risky                    |


I’ll be concentrating on handling the dataset as individual records or editions. Duplicate titles alone are not reliable indicators of duplicate books because different books can share the same or similar titles.


## 8. Cleaning Decision Table

| Priority | Columns                                                   |
| -------- | --------------------------------------------------------- |
| High     | duplicates, authors, genres, ratings, dates, descriptions |
| Medium   | ISBN, pages, language, bookFormat, awards                 |
| Low      | price, characters, setting, edition                       |

## 9. Fake Missing / Placeholder Value Audit

The previous missingness audit only checks formal null values. However, some fields may contain placeholder values, empty list strings, invalid numbers, or text labels that represent missing or unusable information. This section checks for values that are technically present but may need to be treated as missing, invalid, or low-quality during cleaning.

### 9.1 Empty strings and whitespace

In [393]:
str_columns = ['title', 'author', 'publisher', 'description', 'language', 'bookFormat', 'edition', 'series', 'publishDate', 'firstPublishDate']

empty_or_whitespace = df_raw[str_columns].apply(
    lambda col: col.astype(str).str.strip().eq("")
)

empty_or_whitespace.sum()

title               0
author              0
publisher           0
description         0
language            0
bookFormat          0
edition             0
series              1
publishDate         0
firstPublishDate    0
dtype: int64

### 9.2 Placeholders

In [394]:
placeholder_columns = ['language', 'publisher', 'bookFormat', 'edition', 'series', 'description', 'publishDate', 'firstPublishDate']
placeholder_values = ['unknown', 'Unknown', 'UNKNOWN', 'none', 'None', 'N/A', 'n/a', 'na', 'null', 'Null', 'missing', 'not available', 'not specified', '-', '--', '?']
placeholder_values_lower = [v.lower() for v in placeholder_values]

placeholder_mask = df_raw[placeholder_columns].apply(
    lambda col: col.astype(str).str.strip().str.lower().isin(placeholder_values_lower)
)
print(placeholder_mask.sum())

language            0
publisher           1
bookFormat          0
edition             0
series              0
description         0
publishDate         0
firstPublishDate    0
dtype: int64


### 9.3 Empty list-like values

In [395]:
list_columns = ['genres','characters','awards','setting','ratingsByStars']
empty_values = ['[]', '"[]"', '[ ]', '"['']"']
empty_values_lower = [v.lower() for v in empty_values]

empty_mask = df_raw[list_columns].apply(
    lambda col: col.astype(str).str.strip().str.lower().isin(empty_values_lower)
)
print(empty_mask.sum())

genres             4623
characters        38712
awards            41864
setting           40900
ratingsByStars     1423
dtype: int64


### 9.4 Fake ISBN values

In [396]:
s = df_raw["isbn"].astype("string").str.strip().str.lower()
empty_like = s.isin({"", "nan", "none", "null", "na", "n/a"})
all_9s = s.eq("9999999999999")
all_0s = s.str.fullmatch(r"0+").fillna(False)

valid_length = s.str.len().isin([10, 13])
valid_pattern = s.str.fullmatch(r"\d{9}[\dx]|\d{13}").fillna(False)
repeated_suspiciously = s.map(s.value_counts()).ge(10) & ~empty_like

placeholder_like = empty_like | all_9s | all_0s
invalid_length = ~valid_length & ~placeholder_like
invalid_pattern = ~valid_pattern & valid_length & ~placeholder_like
repeated_suspiciously = repeated_suspiciously & ~placeholder_like

print("Placeholder like count:", placeholder_like.sum())
print("Invalid length count:", invalid_length.sum())
print("Invalid pattern count:", invalid_pattern.sum())
print("Repeated suspiciously count:", repeated_suspiciously.sum())


Placeholder like count: 4354
Invalid length count: 28
Invalid pattern count: 4699
Repeated suspiciously count: 0


Although `isbn` has low missingness, it still needs validation because placeholder-like or malformed values are present. ISBN should therefore be treated as a quality-controlled identifier.

### 9.5 Numeric placeholders and impossible values

In [397]:
rating_check = df_raw['rating']

lower_than_zero = rating_check < 0
higher_than_five = rating_check > 5
equal_to_zero = rating_check == 0

print("Lower than 0:", lower_than_zero.sum())
print("Higher than 5:", higher_than_five.sum())
print("Equal to 0:", equal_to_zero.sum())

Lower than 0: 0
Higher than 5: 0
Equal to 0: 71


In [398]:
ratings_check = df_raw['numRatings']

lower_than_zero_ratings = ratings_check < 0
equal_to_zero_ratings = ratings_check == 0

print("Lower than 0:", lower_than_zero_ratings.sum())
print("Equal to 0:", equal_to_zero_ratings.sum())

Lower than 0: 0
Equal to 0: 71


In [399]:
percentage_check = df_raw['likedPercent']

lower_than_zero_percent = percentage_check < 0
over_hundred_percent = percentage_check > 100

print("Lower than 0%:", lower_than_zero_percent.sum())
print("Over 100%:", over_hundred_percent.sum())

Lower than 0%: 0
Over 100%: 0


In [400]:
bbeScore = df_raw['bbeScore']

lower_than_zero_score = bbeScore < 0

print("Lower than 0:", lower_than_zero_score.sum())

Lower than 0: 0


In [401]:
bbeVotes = df_raw['bbeVotes']

lower_than_zero_votes = bbeVotes < 0

print("Lower than 0:", lower_than_zero_votes.sum())

Lower than 0: 52


In [402]:
pages = pd.to_numeric(df_raw['pages'], errors='coerce')

lower_than_zero_pages = pages < 0
equal_to_zero_pages = pages == 0
non_number = pages.isna().sum()

print("Lower than 0:", lower_than_zero_pages.sum())
print("Equal to 0:", equal_to_zero_pages.sum())
print("Non-numeric values:", non_number)

Lower than 0: 0
Equal to 0: 164
Non-numeric values: 2370


In [403]:
price = pd.to_numeric(df_raw['price'], errors='coerce')

lower_than_zero_price = price < 0
equal_to_zero_price = price == 0
non_number_price  = price.isna().sum()

print("Lower than 0:", lower_than_zero_price.sum())
print("Equal to 0:", equal_to_zero_price.sum())
print("Non-numeric values:", non_number_price)

Lower than 0: 0
Equal to 0: 0
Non-numeric values: 14377


Conversion failures must be separated for both `pages` and `price` from original missingness before deciding how many non-numeric values are there.

### 9.6 Date placeholders and bad date strings

In [404]:
publish_date = df_raw['publishDate']
first_publish_date = df_raw['firstPublishDate']

In [405]:
publish_date.head(20)

0     09/14/08
1     09/28/04
2     05/23/06
3     10/10/00
4     09/06/06
5     03/14/06
6     04/28/96
7     09/16/02
8     09/25/12
9     04/01/99
10    01/10/12
11    06/23/07
12    10/07/64
13    10/28/02
14    03/28/06
15    11/22/05
16    06/01/04
17    12/01/00
18    02/04/03
19    03/03/87
Name: publishDate, dtype: str

In [406]:
publish_date.tail(20)

52458      December 1st 1980
52459       January 9th 1997
52460         April 2nd 1998
52461    September 19th 2014
52462                   2004
52463       October 1st 2015
52464     December 12th 2017
52465       January 1st 1983
52466         April 4th 2011
52467     September 4th 2018
52468      January 12th 2017
52469       January 1st 1984
52470       October 1st 2011
52471       August 21st 2011
52472        April 11th 2013
52473          May 28th 2011
52474        August 5th 2011
52475        March 18th 2011
52476     September 1st 2011
52477           May 8th 2011
Name: publishDate, dtype: str

`publishDate` column contains some empty cells, cells with two dates, full date and only year is present in some cells.

In [407]:
first_publish_date.head(10)

0         NaN
1    06/21/03
2    07/11/60
3    01/28/13
4    10/05/05
5    09/01/05
6    08/17/45
7    10/28/56
8    10/20/55
9    06/30/36
Name: firstPublishDate, dtype: str

In [408]:
first_publish_date.tail(20)

52458                   1966
52459                   1994
52460                    NaN
52461                    NaN
52462       January 1st 2002
52463                    NaN
52464                    NaN
52465                    NaN
52466       January 1st 2011
52467                    NaN
52468                    NaN
52469           May 1st 1976
52470    September 16th 2011
52471        August 1st 2011
52472        March 30th 2013
52473                    NaN
52474        August 3rd 2011
52475        March 15th 2011
52476         April 5th 2011
52477                    NaN
Name: firstPublishDate, dtype: str

Because `firstPublishDate` has high missingness, it should be used cautiously. It may still be useful for temporal analysis where available, but missingness should be preserved and documented.

### 9.7 Description quality

In [409]:
description_str = df_raw['description'].str.strip().fillna("")

empty_description = description_str.eq("")
placeholder_description = description_str.str.lower().isin(placeholder_values_lower) 
short_description = description_str.str.len().le(50)
repeated_description = description_str.duplicated(keep=False)
english_description_count = (df_raw.loc[df_raw["language"].str.strip().str.lower() == "english","description"].notna().sum())

print("Empty Description Count:", empty_description.sum())
print("Placeholder Description Count:", placeholder_description.sum())
print("Short Description Count:", short_description.sum())
print("Repeated Description Count:", repeated_description.sum())
print("English Description Count:", english_description_count)

Empty Description Count: 1338
Placeholder Description Count: 0
Short Description Count: 1670
Repeated Description Count: 1726
English Description Count: 42151


For NLP, the description column needs a quality audit beyond missingness. Very short, duplicated, placeholder, or non-target-language descriptions may reduce embedding and clustering quality. Most of the descriptions are in English (42151 out of 52478) with about 3% missingness.

### 9.8 Decision

### Fake Missing Audit Conclusion

The formal missingness audit identifies null values, but several columns contain values that are technically present while still being analytically missing, invalid, or low-quality. These include empty list strings, malformed identifiers, impossible numeric values, and unusable text descriptions.

These values should not all be removed immediately. Instead, they should be inspected, and handled according to the role of each column.

| Column group     | Fake missing / placeholder risk                  | Action                        |
| ---------------- | ------------------------------------------------ | ----------------------------- |
| List-like fields | Empty `[]` values                                | Count and decide per column   |
| ISBN             | Placeholder or malformed identifiers             | Validate before using         |
| Dates            | Different date formats, ambiguous years | Extract year and quality flag |
| Numeric fields   | Negative and null values                       | Flag or clean before EDA      |
| Description      | Empty, short, duplicated, noisy text             | Audit before NLP              |

## 10. Validity Checks Before Cleaning

Before applying cleaning transformations, I check whether key columns contain values that are valid, plausible, and internally consistent. These checks help identify impossible values, suspicious outliers, parsing risks, and contradictions between related fields.


### 10.1 Rating and popularity validity

Rating-related columns are central to the success analysis, so they need stricter validity checks. Average ratings should fall between 0 and 5, liked percentages should fall between 0 and 100, and rating counts should be non-negative. The `ratingsByStars` field also needs structural validation because it should contain five star-count values.

In [410]:
rating_check = df_raw['rating']

lower_than_zero = rating_check < 0
higher_than_five = rating_check > 5

print("Lower than 0:", lower_than_zero.sum())
print("Higher than 5:", higher_than_five.sum())

Lower than 0: 0
Higher than 5: 0


In [411]:
percentage_check = df_raw['likedPercent']

lower_than_zero_percent = percentage_check < 0
over_hundred_percent = percentage_check > 100

print("Lower than 0%:", lower_than_zero_percent.sum())
print("Over 100%:", over_hundred_percent.sum())

Lower than 0%: 0
Over 100%: 0


In [412]:
ratings_check = df_raw['numRatings']

negative_ratings= ratings_check < 0

print("Lower than 0:", negative_ratings.sum())

Lower than 0: 0


In [413]:
def has_5_elements(value):
    try:
        parsed = ast.literal_eval(value)
        return isinstance(parsed, list) and len(parsed) == 5
    except (ValueError, SyntaxError, TypeError):
        return False
    
def has_0_elements(value):
    try:
        parsed = ast.literal_eval(value)
        return isinstance(parsed, list) and len(parsed) == 0
    except (ValueError, SyntaxError, TypeError):
        return False    

In [414]:
def sum_list_string(value):
    try:
        parsed = ast.literal_eval(value)

        if not isinstance(parsed, list):
            return None

        return sum(float(x) for x in parsed)

    except (ValueError, SyntaxError, TypeError):
        return None

In [415]:
ratings_count_check = df_raw['ratingsByStars'].apply(has_5_elements)

In [416]:
null_ratings_count_check = df_raw['ratingsByStars'].apply(has_0_elements)

In [417]:
sum_star_ratings = df_raw["ratingsByStars"].apply(sum_list_string)
matches_numRatings = sum_star_ratings == df_raw['numRatings']

In [418]:
print("ratingsByStars with 5 elements:",ratings_count_check.sum())
print("ratingsByStars with 0 elements:",null_ratings_count_check.sum())
print("ratingsByStars matches numRatings:",matches_numRatings.sum())

ratingsByStars with 5 elements: 51055
ratingsByStars with 0 elements: 1423
ratingsByStars matches numRatings: 51126


Rows with empty `ratingsByStar` appear to be mostly low-engagement books, often with very low `numRatings` and `bbeVotes`. Some empty `ratingsByStars` rows also have `numRatings` == 0, which explains why a small number of empty lists still match `numRatings`.

### 10.2 Popularity / ranking validity

Popularity and ranking columns should be checked for impossible or suspicious values before analysis. Negative vote counts are especially important because they may indicate a scraping issue, placeholder value, or dataset-specific artifact.

Most negative votes or zero ratings rows have a lot of missing values, which could indicate scraping issues.

In [419]:
bbeScore = df_raw['bbeScore']

lower_than_zero_score = bbeScore < 0

print("Lower than 0:", lower_than_zero_score.sum())

Lower than 0: 0


In [420]:
bbeVotes = df_raw['bbeVotes']

lower_than_zero_votes = bbeVotes < 0

print("Lower than 0:", lower_than_zero_votes.sum())

Lower than 0: 52


In [421]:
numRatings = df_raw['numRatings']

lower_than_zero_ratings = numRatings < 0

print("Lower than 0:", lower_than_zero_ratings.sum())

Lower than 0: 0


### 10.3 Page count validity

In [422]:
pages = pd.to_numeric(df_raw['pages'].astype(str).str.extract(r'(\d+)', expand=False),errors='coerce')

zero_pages = pages == 0
negative_pages = pages < 0
high_pages = pages > 1000
blank_pages = pages.isna()

print('Zero pages: ', zero_pages.sum())
print('Negative pages: ', negative_pages.sum())
print('High page count : ', high_pages.sum())
print('Blank page count: ', blank_pages.sum())

Zero pages:  164
Negative pages:  0
High page count :  651
Blank page count:  2347


Zero-page values are suspicious and should be treated as invalid or placeholder-like unless there is a clear reason to keep them. Very high page counts should be inspected because they may represent collections, box sets, or parsing issues.

### 10.4 Price validity

In [423]:
price = pd.to_numeric(df_raw['price'].astype(str).str.extract(r'(\d+(?:\.\d+)?)', expand=False),errors='coerce')

zero_price = price == 0
negative_price = price < 0
high_price = price > 50
no_price = price.isna()

print('Zero price: ', zero_price.sum())
print('Negative price: ', negative_price.sum())
print('High price: ', high_price.sum())
print('No price count: ', no_price.sum())

Zero price:  0
Negative price:  0
High price:  742
No price count:  14365


Price has substantial missingness and some high values, so it should remain a secondary feature unless price analysis becomes part of the final project. High prices should be inspected before deciding whether they are true outliers or valid expensive editions/sets.

### 10.5 Date validity

In [424]:
missing = (publish_date.isna() | publish_date.astype(str).str.strip().eq("") | publish_date.astype(str).str.strip().eq("Unknown") ).sum()

year_only_pattern = r"^\d{4}$"
year_only = publish_date.astype(str).str.strip().str.fullmatch(year_only_pattern).sum()

month_year_text_pattern = r"^(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{4}$"
month_year_text = publish_date.astype(str).str.strip().str.fullmatch(month_year_text_pattern).sum()

month_day_year_text_pattern = r"^(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+(?:(?:1|21|31)st|(?:2|22)nd|(?:3|23)rd|(?:[4-9]|1[0-9]|20|2[4-9]|30)th)(?:,\s*)?\s+\d{4}$"
month_day_year_text = publish_date.astype(str).str.strip().str.fullmatch(month_day_year_text_pattern).sum()

two_number_year_pattern = r"\b(?:0?[1-9]|1[0-2])/(?:0?[1-9]|[12][0-9]|3[01])/\d{2}\b"
numeric_mmddyy = publish_date.astype(str).str.strip().str.fullmatch(two_number_year_pattern).sum()

four_number_year_pattern = r"\b(?:0?[1-9]|1[0-2])/(?:0?[1-9]|[12][0-9]|3[01])/\d{4}\b"
numeric_mmddyyyy = publish_date.astype(str).str.strip().str.fullmatch(four_number_year_pattern).sum()

unknown_format = len(publish_date) - (missing + year_only + month_year_text + month_day_year_text + numeric_mmddyy + numeric_mmddyyyy)

print('Missing dates: ', missing)
print('Unknown format dates: ', unknown_format)
print('Year only dates: ', year_only)
print('Month-year text dates: ', month_year_text)
print('Month-day-year text dates: ', month_day_year_text)
print('Numeric MM/DD/YY dates ', numeric_mmddyy)
print('Numeric MM/DD/YYYY dates: ', numeric_mmddyyyy)

Missing dates:  880
Unknown format dates:  905
Year only dates:  4362
Month-year text dates:  2199
Month-day-year text dates:  43314
Numeric MM/DD/YY dates  818
Numeric MM/DD/YYYY dates:  0


In [425]:
publish_date_clean = publish_date.astype(str).str.strip()

known_format_mask = (
    publish_date_clean.eq("") |
    publish_date_clean.eq("Unknown") |
    publish_date_clean.isna()|
    publish_date_clean.str.fullmatch(year_only_pattern).fillna(False) |
    publish_date_clean.str.fullmatch(month_year_text_pattern).fillna(False) |
    publish_date_clean.str.fullmatch(month_day_year_text_pattern).fillna(False) |
    publish_date_clean.str.fullmatch(two_number_year_pattern).fillna(False) |
    publish_date_clean.str.fullmatch(four_number_year_pattern).fillna(False)
)

unknown_mask = ~known_format_mask
unknown_format = unknown_mask.sum()

print("Unknown format sample values:")
print(publish_date_clean[unknown_mask].head(20))

Unknown format sample values:
1623                                            Published
2271                                            Published
2489                                            Published
2989    Best Books to Read When the Snow Is Falling\n\...
3145                                            Published
3560    Most Interesting World\n\n3,055 books — 2,474 ...
3943                                         April 18th 1
4006                                            Published
4163                                            Published
4332                                            Published
4358                                            Published
4477                                            Published
4508    في أحضان الكتب - الجزء الثاني\n\n111 books — 2...
4631                                            Published
4966                                            Published
5050              Expected publication: October 3rd 20136
5488                                      

The unknown format sample contains book suggestion lists, the word 'Published', incorrectly formatted dates.

### Example values

| Category              | Example values                          |
| --------------------- | --------------------------------------- |
| `missing`             | blank / NaN                             |
| `year_only`           | `2004`, `1966`                          |
| `month_year_text`     | `March 2005`                            |
| `month_day_year_text` | `December 1st 1980`, `January 9th 1997` |
| `numeric_mmddyy`      | `09/14/08`, `06/21/03`                  |

Numeric dates with two-digit years are ambiguous and should be flagged before cleaning. They may require century inference using book context, title, or `firstPublishDate`.

In [426]:
missing = (first_publish_date.isna() | first_publish_date.astype(str).str.strip().eq("") | first_publish_date.astype(str).str.strip().eq("Unknown")).sum()

year_only_pattern = r"^\d{4}$"
year_only = first_publish_date.astype(str).str.strip().str.fullmatch(year_only_pattern).sum()

month_year_text_pattern = r"^(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{4}$"
month_year_text = first_publish_date.astype(str).str.strip().str.fullmatch(month_year_text_pattern).sum()

month_day_year_text_pattern = r"^(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+(?:(?:1|21|31)st|(?:2|22)nd|(?:3|23)rd|(?:[4-9]|1[0-9]|20|2[4-9]|30)th)(?:,\s*)?\s+\d{4}$"
month_day_year_text = first_publish_date.astype(str).str.strip().str.fullmatch(month_day_year_text_pattern).sum()

two_number_year_pattern = r"\b(?:0?[1-9]|1[0-2])/(?:0?[1-9]|[12][0-9]|3[01])/\d{2}\b"
numeric_mmddyy = first_publish_date.astype(str).str.strip().str.fullmatch(two_number_year_pattern).sum()

four_number_year_pattern = r"\b(?:0?[1-9]|1[0-2])/(?:0?[1-9]|[12][0-9]|3[01])/\d{4}\b"
numeric_mmddyyyy = first_publish_date.astype(str).str.strip().str.fullmatch(four_number_year_pattern).sum()

unknown_format = len(first_publish_date) - (missing + year_only + month_year_text + month_day_year_text + numeric_mmddyy + numeric_mmddyyyy)

print('Missing dates: ', missing)
print('Unknown format dates: ', unknown_format)
print('Year only dates: ', year_only)
print('Month-year text dates: ', month_year_text)
print('Month-day-year text dates: ', month_day_year_text)
print('Numeric MM/DD/YY dates ', numeric_mmddyy)
print('Numeric MM/DD/YYYY dates: ', numeric_mmddyyyy)

Missing dates:  21326
Unknown format dates:  4
Year only dates:  1031
Month-year text dates:  151
Month-day-year text dates:  1186
Numeric MM/DD/YY dates  28780
Numeric MM/DD/YYYY dates:  0


In [427]:
first_publish_date_clean = first_publish_date.astype(str).str.strip()

known_format_mask = (
    first_publish_date_clean.eq("") |
    first_publish_date_clean.eq("Unknown") |
    first_publish_date_clean.isna() |
    first_publish_date_clean.str.fullmatch(year_only_pattern).fillna(False) |
    first_publish_date_clean.str.fullmatch(month_year_text_pattern).fillna(False) |
    first_publish_date_clean.str.fullmatch(month_day_year_text_pattern).fillna(False) |
    first_publish_date_clean.str.fullmatch(two_number_year_pattern).fillna(False) |
    first_publish_date_clean.str.fullmatch(four_number_year_pattern).fillna(False)
)

unknown_mask = ~known_format_mask
unknown_format = unknown_mask.sum()

print("Unknown format sample values:")
print(first_publish_date_clean[unknown_mask].head(20))

Unknown format sample values:
48937    14th 1857
50114           61
52224          327
52280         -458
Name: firstPublishDate, dtype: str


In [428]:
df_raw.iloc[[48937, 50114, 52224, 52280]]

,bookId,title,series,author,rating,description,language,isbn,genres,characters,...,firstPublishDate,awards,numRatings,ratingsByStars,likedPercent,setting,coverImg,bbeScore,bbeVotes,price
48937,6098820-the-life-of-st-dominic,The Life of St. Dominic,NaN,Augusta Theodosia Drane,4.39,There are few biographies of St. Dominic that ...,English,9780895553362,"['Biography', 'Religion', 'Catholic']",[],...,14th 1857,[],23,"['11', '10', '2', '0', '0']",100.0,[],https://i.gr-assets.com/images/S/compressed.ph...,46,1,11.42
50114,11821471-civil-war,Civil War,NaN,"Marcus Annaeus Lucanus, Matthew Fox (Translato...",3.93,A magnificent new translation of the enduring ...,English,9780143106234,"['Classics', 'Poetry', 'History', 'Roman', 'An...",[],...,61,[],1211,"['434', '368', '312', '79', '18']",92.0,[],https://i.gr-assets.com/images/S/compressed.ph...,36,1,6.25
52224,25063232-die-sch-nsten-erz-hlungen-2,Die schönsten Erzählungen 2,NaN,Jeremias Gotthelf,5.00,NaN,German,9999999999999,[],[],...,327,[],1,[],NaN,[],NaN,6,1,NaN
52280,1524.Agamemnon,Agamemnon,Oresteia #1,"Aeschylus, Judith Affleck, Philip de May (Edit...",3.85,Treating ancient plays as living drama. Classi...,English,9780521010757,"['Classics', 'Plays', 'Drama', 'Mythology', 'F...","['Cassandra (of Troy)', 'Clytemnestra (wife of...",...,-458,[],14958,"['4299', '5340', '4232', '906', '181']",93.0,[],https://i.gr-assets.com/images/S/compressed.ph...,5,1,0.86


#### Example values

| Category              | Example values                          |
| --------------------- | --------------------------------------- |
| `missing`             | blank / NaN                             |
| `year_only`           | `2004`, `1966`                          |
| `month_year_text`     | `March 2005`                            |
| `month_day_year_text` | `December 1st 1980`, `January 9th 1997` |
| `numeric_mmddyy`      | `09/14/08`, `06/21/03`                  |

The remaining unknown `firstPublishDate` values are not ordinary formatting errors. They include one malformed partial date, two short historical year values, and one negative year that likely represents BCE. These should not be force-parsed with the main date patterns.

For the cleaning stage, these values should be handled separately from standard date formats. Some may still contain useful year information, but they require a special quality flag because they are not directly comparable to normal modern publication dates without interpretation.

In [429]:
def parse_year_and_flag(date_series):
    clean = date_series.fillna("").astype(str).str.strip()
    placeholder_set = set(placeholder_values_lower)

    missing_mask = date_series.isna() | clean.eq("") | clean.str.lower().isin(placeholder_set) | clean.str.contains("Published", na=False) | clean.str.contains("Books", na=False) | clean.str.contains("books", na=False) | clean.str.contains("Expected publication:", na=False)
    full_date_month_mask = clean.str.fullmatch(month_day_year_text_pattern).fillna(False)
    full_date_numeric_mask = clean.str.fullmatch(four_number_year_pattern).fillna(False)
    partial_month_mask = clean.str.fullmatch(month_year_text_pattern).fillna(False)
    year_only_mask = clean.str.fullmatch(year_only_pattern).fillna(False)
    ambiguous_two_digit_mask = clean.str.fullmatch(two_number_year_pattern).fillna(False)

    quality = pd.Series("unparseable", index=date_series.index, dtype="string")
    quality[missing_mask] = "missing"
    quality[full_date_month_mask | full_date_numeric_mask] = "full_date_reliable"
    quality[partial_month_mask] = "partial_month_year"
    quality[year_only_mask] = "partial_year_only"
    quality[ambiguous_two_digit_mask] = "ambiguous_two_digit_year"

    year = pd.Series(pd.NA, index=date_series.index, dtype="Int64")

    month_day_year_extract = (
        r"^(?:January|February|March|April|May|June|July|August|September|October|November|December)"
        r"\s+(?:(?:1|21|31)st|(?:2|22)nd|(?:3|23)rd|(?:[4-9]|1[0-9]|20|2[4-9]|30)th)"
        r"(?:,\s*)?\s+(\d{4})$"
    )
    month_year_extract = r"^(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+(\d{4})$"
    four_digit_extract = r"^(?:0?[1-9]|1[0-2])/(?:0?[1-9]|[12][0-9]|3[01])/(\d{4})$"
    year_only_extract = r"^(\d{4})$"

    year.loc[full_date_month_mask] = clean.loc[full_date_month_mask].str.extract(month_day_year_extract, expand=False).astype("Int64")
    year.loc[full_date_numeric_mask] = clean.loc[full_date_numeric_mask].str.extract(four_digit_extract, expand=False).astype("Int64")
    year.loc[partial_month_mask] = clean.loc[partial_month_mask].str.extract(month_year_extract, expand=False).astype("Int64")
    year.loc[year_only_mask] = clean.loc[year_only_mask].str.extract(year_only_extract, expand=False).astype("Int64")
    year.loc[ambiguous_two_digit_mask] = pd.NA

    return year, quality

df_raw["publish_year"], df_raw["publish_year_quality"] = parse_year_and_flag(df_raw["publishDate"])
df_raw["first_publish_year"], df_raw["first_publish_year_quality"] = parse_year_and_flag(df_raw["firstPublishDate"])

In [430]:

df_raw["first_publish_year_quality"].eq("full_date_reliable").sum()


np.int64(1186)

#### Date Validity Summary

| Quality flag               | `publishDate` count | `firstPublishDate` count |
| -------------------------- | ------------------: | -----------------------: |
| `full_date_reliable`       |               43314 |                     1186 |
| `partial_year_only`        |                4362 |                     1031 |
| `partial_month_year`       |                2199 |                      151 |
| `ambiguous_two_digit_year` |                 818 |                    28780 |
| `missing`                  |                 1771 |                    21326 |
| `unparseable`              |                14 |                        4 |


In [431]:
df_raw.loc[
    df_raw["publish_year_quality"].eq("unparseable"),
    ["publishDate", "publish_year", "publish_year_quality"]
].head(14)

,publishDate,publish_year,publish_year_quality
3943,April 18th 1,<NA>,unparseable
9491,Great Erotica,<NA>,unparseable
11231,September 9th 214,<NA>,unparseable
20431,Heroine wrongly accused,<NA>,unparseable
23436,March 30th 17,<NA>,unparseable
23517,LitHub's Best of The Decade,<NA>,unparseable
23696,Comfort Reads for Tweens,<NA>,unparseable
24586,Bought,<NA>,unparseable
25468,January 23,<NA>,unparseable
27792,مائة,<NA>,unparseable


#### Date Validity Conclusion

The date columns contain mixed levels of completion and quality. Most `publishDate` values are usable for year-level analysis, with the majority classified as `full_date_reliable`, `partial_year_only`, or `partial_month_year`.

The `firstPublishDate` column is less complete. A large amount of values are either missing or use ambiguous two-digit years. These two-digit years are not converted into full years during the audit because automatic parsing could assign the wrong century, especially for older books.

### 10.6 ISBN Validity

This section will check whether ISBN can be used as an identifier, or does it contain placeholder or malformed values that could create false positives down the line during the project.

#### ISBN structure counts

In [432]:
isbn = df_raw["isbn"].fillna("").astype(str).str.strip()

isbn_normalized = (
    isbn
    .str.upper()
    .str.replace("-", "", regex=False)
    .str.replace(" ", "", regex=False)
)

missing_like_values = {"", "NAN", "NONE", "NULL", "NA", "N/A"}
missing_mask = isbn_normalized.isin(missing_like_values)

length = isbn_normalized.str.len()

isbn_like_pattern = r"(?:\d{9}[\dX]|\d{13})"
isbn_like_mask = isbn_normalized.str.fullmatch(isbn_like_pattern).fillna(False)

placeholder_mask = (
    isbn_normalized.eq("9999999999999")
    | isbn_normalized.str.fullmatch(r"0+").fillna(False)
)

invalid_length_mask = (
    ~missing_mask
    & ~placeholder_mask
    & ~length.isin([10, 13])
)

invalid_characters_mask = (
    ~missing_mask
    & ~placeholder_mask
    & length.isin([10, 13])
    & ~isbn_like_mask
)

repeated_threshold = 10

repeat_base_mask = (
    ~missing_mask
    & ~placeholder_mask
    & isbn_like_mask
)

isbn_counts = isbn_normalized.where(repeat_base_mask).value_counts()

repeated_count = isbn_normalized.map(isbn_counts)

repeated_suspicious_mask = (
    repeat_base_mask
    & repeated_count.ge(repeated_threshold).fillna(False)
)

print("ISBN-like structure:", isbn_like_mask.sum())
print("Missing-like:", missing_mask.sum())
print("Placeholder-like:", placeholder_mask.sum())
print("Invalid length:", invalid_length_mask.sum())
print("Invalid characters with valid length:", invalid_characters_mask.sum())
print("Repeated suspicious:", repeated_suspicious_mask.sum())


ISBN-like structure: 47751
Missing-like: 0
Placeholder-like: 4354
Invalid length: 28
Invalid characters with valid length: 4699
Repeated suspicious: 0


| ISBN status                 | Count |
| --------------------------- | ----: |
| ISBN-like structure                 |   47751 |
| Missing-like |   0 |
| Placeholder-like             |   4354 |
| Invalid length                     |   28 |
| Invalid characters with valid length  |   4699 |
| Repeated suspicious excluding placeholders                      |   0 |

In [433]:
isbn_normalized.value_counts().head()

isbn
9999999999999    4354
9780765326355       2
9780312429980       2
9781250166548       2
9780764208096       2
Name: count, dtype: int64

In [434]:
df_raw.loc[
    isbn_normalized.eq("9999999999999"),
    ["title", "author", "isbn", "bookFormat", "publishDate"]
].head(10)

,title,author,isbn,bookFormat,publishDate
2,To Kill a Mockingbird,Harper Lee,9999999999999,Paperback,05/23/06
3,Pride and Prejudice,"Jane Austen, Anna Quindlen (Introduction)",9999999999999,Paperback,10/10/00
7,The Chronicles of Narnia,"C.S. Lewis, Pauline Baynes (Illustrator)",9999999999999,Paperback,09/16/02
10,The Fault in Our Stars,John Green (Goodreads Author),9999999999999,Hardcover,01/10/12
11,The Hitchhiker's Guide to the Galaxy,Douglas Adams,9999999999999,Paperback,06/23/07
14,The Da Vinci Code,Dan Brown (Goodreads Author),9999999999999,Paperback,03/28/06
16,The Picture of Dorian Gray,"Oscar Wilde, Jeffrey Eugenides (Introduction)",9999999999999,Paperback,06/01/04
19,Les Misérables,"Victor Hugo, Lee Fahnestock (Translator), Norm...",9999999999999,Mass Market Paperback,03/03/87
26,The Perks of Being a Wallflower,Stephen Chbosky,9999999999999,Paperback,02/28/99
27,The Great Gatsby,"F. Scott Fitzgerald, Francis Scott Fitzgerald",9999999999999,Paperback,09/28/04


Top repeated normalized ISBN value is 9999999999999, this seems to be used as a placeholder value for famous or classical books. No other ISBN is repeated as much as 9999999999999. Repeated ISBN-like values consistently match the same title and author, but a small number have edition-level metadata differences. ISBN is useful as a duplicate signal after excluding placeholders and invalid identifiers, but it should not be the only automatic deletion rule.

The rows/books with ISBN 9999999999999 should stay in the dataset, but placeholder ISBN value should be excluded from ISBN-based duplicate matching. During cleaning create a normalized column and a flag column to show the quality of the ISBN and excluding placeholders and invalid structures.

#### ISBN summary table

In [435]:
valid_for_isbn_duplicate_check = (
    ~missing_mask
    & ~placeholder_mask
    & isbn_like_mask
)

isbn_duplicate_base = df_raw.loc[
    valid_for_isbn_duplicate_check,
    ["title", "author", "isbn", "bookFormat", "publishDate", "publisher", "language"]
].copy()

isbn_duplicate_base["isbn_normalized"] = isbn_normalized[valid_for_isbn_duplicate_check]

isbn_group_summary = (
    isbn_duplicate_base
    .groupby("isbn_normalized")
    .agg(
        rows=("isbn_normalized", "size"),
        unique_titles=("title", lambda x: x.nunique(dropna=False)),
        unique_authors=("author", lambda x: x.nunique(dropna=False)),
        unique_formats=("bookFormat", lambda x: x.nunique(dropna=False)),
        unique_publish_dates=("publishDate", lambda x: x.nunique(dropna=False)),
        unique_publishers=("publisher", lambda x: x.nunique(dropna=False)),
        unique_languages=("language", lambda x: x.nunique(dropna=False)),
    )
    .reset_index()
)

duplicated_isbn_groups = isbn_group_summary.query("rows > 1")

In [436]:
isbn_duplicate_metrics = pd.Series({
    "duplicated non-placeholder ISBN groups": duplicated_isbn_groups.shape[0],

    "repeated ISBN rows": duplicated_isbn_groups["rows"].sum(),

    "groups with same title + author": (
        (duplicated_isbn_groups["unique_titles"] == 1)
        & (duplicated_isbn_groups["unique_authors"] == 1)
    ).sum(),

    "groups with differing format/date/publisher/language": (
        (duplicated_isbn_groups["unique_formats"] > 1)
        | (duplicated_isbn_groups["unique_publish_dates"] > 1)
        | (duplicated_isbn_groups["unique_publishers"] > 1)
        | (duplicated_isbn_groups["unique_languages"] > 1)
    ).sum(),

    "groups with differing title or author": (
        (duplicated_isbn_groups["unique_titles"] > 1)
        | (duplicated_isbn_groups["unique_authors"] > 1)
    ).sum(),
})

isbn_duplicate_metrics

duplicated non-placeholder ISBN groups                  45
repeated ISBN rows                                      90
groups with same title + author                         45
groups with differing format/date/publisher/language     3
groups with differing title or author                    0
dtype: int64

### 10.7 Genres validity

This section will cover the validity checks for genres. Genres column is central to the analysis, empty or malformed genre lists would limit filtering and group comparisons.

In [437]:
def parse_list_cell(value):
    if pd.isna(value):
        return [], False
    try:
        parsed = ast.literal_eval(value)
        if isinstance(parsed, list):
            return parsed, False
        return [], True
    except (ValueError, SyntaxError):
        return [], True

genre_parse= df_raw["genres"].apply(parse_list_cell)

parsed_genres = genre_parse.apply(lambda x: x[0])
genre_parse_failed = genre_parse.apply(lambda x: x[1])

genre_parse_failed.sum()

np.int64(0)

In [438]:
sum(parsed_genres.apply(lambda x: len(x) == 0))

4623

In [439]:
sum(parsed_genres.apply(lambda x: len(x) >= 1))

47855

Genres values are parseable list strings. 4623 rows have empty genre lists, however no rows fail parsing. 47855 have at least one genre.

In [440]:
all_genres = [genre for genres in parsed_genres for genre in genres]
genre_counts = pd.Series(Counter(all_genres)).sort_values(ascending=False)
genre_counts.head(20)


Fiction               31638
Romance               15495
Fantasy               15046
Young Adult           11869
Contemporary          10520
Nonfiction             8251
Adult                  8246
Novels                 7805
Mystery                7702
Historical Fiction     7665
Audiobook              7307
Classics               6902
Adventure              6452
Historical             6383
Paranormal             6030
Literature             5836
Science Fiction        5374
Childrens              5226
Thriller               4587
Magic                  4248
dtype: int64

Genre space is quite granular with a long tail towards the end. Top 20 genres cover only 45% of genre assignments, hence many labels outside the top genres. Many top genres and genre families are overlapping.

In [441]:
genre_counts.shape[0]

982

In [442]:
rare_genres = genre_counts[genre_counts < 10]

rare_genres.shape[0]

317

In [443]:
genre_counts.head(20).sum() / genre_counts.sum()

np.float64(0.44781442075159794)

In [444]:
(genre_counts == 1).sum()

np.int64(87)

| Metric                               | Value |
| ------------------------------------ | ----: |
| rows with at least one genre         |   47855 |
| rows with empty genre list           |   4623 |
| unique genre labels                  |   982 |
| genres appearing once                |   87 |
| genres appearing fewer than 10 times |   317 |
| top 20 genre assignment share        |   45% |


In [445]:
genre_count_per_book = parsed_genres.apply(len)

genre_count_per_book.describe()

count    52478.000000
mean         7.769313
std          3.578427
min          0.000000
25%          6.000000
50%         10.000000
75%         10.000000
max         10.000000
Name: genres, dtype: float64

In [446]:
genre_count_per_book.value_counts().sort_index().head(20)

genres
0      4623
1      2070
2      1548
3      1500
4      1518
5      1479
6      1377
7      1181
8      1257
9      1250
10    34675
Name: count, dtype: int64

10 genres might not be a natural pattern but a scraping or GoodReads limit. Genre oder might matter. Later cleaning should use top/primary genre.


### 10.8 Ratings By Stars validity

This section will cover validity checks for `ratingsByStars` and how consistent it is with `numRatings`.

In [447]:
ratings_parse= df_raw["ratingsByStars"].apply(parse_list_cell)

parsed_ratings = ratings_parse.apply(lambda x: x[0])
ratings_parse_failed = ratings_parse.apply(lambda x: x[1])

ratings_parse_failed.sum()

np.int64(0)

In [448]:
def clean_star_value(value):
    try:
        return int(str(value).replace(",", "").strip())
    except (ValueError, TypeError):
        return pd.NA

In [449]:
cleaned_star_counts = parsed_ratings.apply(
    lambda values: [clean_star_value(v) for v in values]
)

In [450]:
cleaned_star_counts.head()

0      [3444695, 1921313, 745221, 171994, 93557]
1        [1593642, 637516, 222366, 39573, 14526]
2      [2363896, 1333153, 573280, 149952, 80794]
3       [1617567, 816659, 373311, 113934, 76770]
4    [1751460, 1113682, 1008686, 542017, 548674]
Name: ratingsByStars, dtype: object

In [451]:
has_non_numeric_star_value = cleaned_star_counts.apply(lambda values: any(pd.isna(v) for v in values))
has_non_numeric_star_value.sum()

np.int64(0)

In [452]:
ratings_by_stars_sum = cleaned_star_counts.apply(lambda values: sum(v for v in values if pd.notna(v)))

In [453]:
rating_count_difference = ratings_by_stars_sum - df_raw["numRatings"]

In [454]:
(rating_count_difference == 0).sum()

np.int64(51126)

In [455]:
(rating_count_difference != 0).sum()

np.int64(1352)

In [456]:
rating_count_difference.describe()

count    52478.000000
mean        -0.259042
std         44.854919
min     -10275.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          0.000000
dtype: float64

In [457]:
rating_count_difference.abs().sort_values(ascending=False).head(20)

26025    10275
33379        4
21392        4
11491        4
38674        4
34856        4
30405        4
23095        4
5003         4
41805        4
21099        4
20624        4
29146        4
23807        4
38669        4
47947        4
25360        4
22583        4
41300        4
29933        4
dtype: int64

In [458]:
ratings_by_stars_missing = (df_raw["ratingsByStars"].isna() | df_raw["ratingsByStars"].eq("") | df_raw["ratingsByStars"].eq("[]"))
ratings_by_stars_missing.sum()

np.int64(1423)

In [459]:
wrong_length = (~ratings_by_stars_missing & cleaned_star_counts.apply(lambda values: len(values) != 5))
non_numeric = cleaned_star_counts.apply(lambda values: any(pd.isna(v) for v in values))
wrong_length.sum(), non_numeric.sum()

(np.int64(0), np.int64(0))

In [460]:
ratings_by_stars_malformed = wrong_length | non_numeric
ratings_by_stars_malformed.sum()

np.int64(0)

In [461]:
difference_abs = rating_count_difference.abs()

In [462]:
relative_difference = difference_abs / df_raw["numRatings"].replace(0, pd.NA)

In [463]:
well_formed_ratings_mask = (~ratings_by_stars_missing & ~wrong_length & ~non_numeric)

In [464]:
ratings_by_stars_inconsistent = (well_formed_ratings_mask & (relative_difference > 0.01))
ratings_by_stars_inconsistent.sum()

np.int64(0)

In [465]:
well_formed_exact_match = (well_formed_ratings_mask & (rating_count_difference == 0))

well_formed_any_difference = (well_formed_ratings_mask & (rating_count_difference != 0))

ratings_by_stars_audit = pd.Series({
    "missing": ratings_by_stars_missing.sum(),
    "wrong list length among non-missing": wrong_length.sum(),
    "contains non-numeric values": non_numeric.sum(),
    "well-formed rows": well_formed_ratings_mask.sum(),
    "well-formed exact sum matches numRatings": well_formed_exact_match.sum(),
    "well-formed sum differs from numRatings": well_formed_any_difference.sum(),
    "well-formed sum differs by more than 1%": ratings_by_stars_inconsistent.sum(),
})

ratings_by_stars_audit

missing                                      1423
wrong list length among non-missing             0
contains non-numeric values                     0
well-formed rows                            51055
well-formed exact sum matches numRatings    51055
well-formed sum differs from numRatings         0
well-formed sum differs by more than 1%         0
dtype: int64

The star-count lists are parseable, but they contain either 5 values or are missing. The values are numeric after parsing them to integers. Summed star counts usually match `numRatings`. The inconsistencies seem rare and the field can still be used. During cleaning it should be split into separate columns for each of the stars for better usability later on in the project.

### 10.9 Characters, awards, and setting validity

This section will cover validity tests for these columns to check if they are usable for analysis or are they too sparse to be useful core features.

In [467]:
characters_parse= df_raw["characters"].apply(parse_list_cell)

parsed_characters = characters_parse.apply(lambda x: x[0])
characters_parse_failed = characters_parse.apply(lambda x: x[1])

characters_parse_failed.sum()

np.int64(0)

In [468]:
awards_parse= df_raw["awards"].apply(parse_list_cell)

parsed_awards = awards_parse.apply(lambda x: x[0])
awards_parse_failed = awards_parse.apply(lambda x: x[1])

awards_parse_failed.sum()

np.int64(0)

In [469]:
setting_parse= df_raw["setting"].apply(parse_list_cell)

parsed_settings = setting_parse.apply(lambda x: x[0])
setting_parse_failed = setting_parse.apply(lambda x: x[1])

setting_parse_failed.sum()

np.int64(0)

In [470]:
characters_missing = (df_raw["characters"].isna() | df_raw["characters"].eq("") | df_raw["characters"].eq("[]"))
characters_missing.sum()

np.int64(38712)

In [471]:
awards_missing = (df_raw["awards"].isna() | df_raw["awards"].eq("") | df_raw["awards"].eq("[]"))
awards_missing.sum()

np.int64(41864)

In [472]:
setting_missing = (df_raw["setting"].isna() | df_raw["setting"].eq("") | df_raw["setting"].eq("[]"))
setting_missing.sum()

np.int64(40900)

In [477]:
summary = pd.DataFrame({
    "awards": {
        "empty list %": round((awards_missing.sum() * 100) / len(df_raw), 0),
        "parse failures": awards_parse_failed.sum(),
        "rows with values": (parsed_awards.apply(len) > 0).sum(),
    },
    "setting": {
        "empty list %": round((setting_missing.sum() * 100) / len(df_raw), 0),
        "parse failures": setting_parse_failed.sum(),
        "rows with values": (parsed_settings.apply(len) > 0).sum(),
    },
    "characters": {
        "empty list %": round((characters_missing.sum() * 100) / len(df_raw), 0),
        "parse failures": characters_parse_failed.sum(),
        "rows with values": (parsed_characters.apply(len) > 0).sum(),
    },
}).T
summary

,empty list %,parse failures,rows with values
awards,80.0,0.0,10614.0
setting,78.0,0.0,11578.0
characters,74.0,0.0,13766.0


`characters`, `awards`, and `setting` are sparse list-like columns. The non-empty are parseable. They should not drive the main analysis, but they may be kept for optional features or niche exploration after parsing. They should be lower priority than `genres` and `ratingsByStars`.

### 10.10 Author and contributor validity

In [ ]:
author_has_parentheses = df_raw['author'].astype(str).str.contains(r"[()]", na=False)
author_has_parentheses.value_counts().head(20)

author
True     29056
False    23422
Name: count, dtype: int64

In [483]:
author_parenthetical_text = (df_raw["author"].dropna().astype(str).str.findall(r"\(([^()]*)\)"))
parenthetical_counts = (author_parenthetical_text.explode().dropna().str.strip().value_counts())

parenthetical_counts.head(30)

author
Goodreads Author    24368
Translator           2896
Illustrator          2337
Editor               1221
Introduction          754
Contributor           673
Illustrations         330
Foreword              271
Narrator              230
Artist                182
Writer                161
Translation           139
translator            130
Foreword by           101
Creator                97
Afterword              83
pseudonym              61
Pseudonym              56
illustrator            49
ترجمة                  49
Colorist               48
Letterer               45
Preface                43
Photographer           43
Reading                36
Adapter                35
editor                 30
Übersetzer             29
Inker                  28
Annotations            27
Name: count, dtype: int64

The `author` column seems to mix three things: platform label, contributor roles, and has inconsistent role spellings and different languages. This means that `author` column should not be used raw for author-level analysis and platform label shoud be removed, contributor roles should be parsed separately.